runtime/parachains/src/dmp/ringbuf.rs

```diff
@@ -0,0 +1,437 @@
+// Copyright 2020 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Implements API for managing a ring buffer and an associated message window.
+
+use frame_support::pallet_prelude::*;
+use polkadot_parachain::primitives::{MessageIndex, PageIndex, WrappingIndex};
+use primitives::v2::Id as ParaId;
+use sp_std::prelude::*;
+
+/// Unique identifier of an inbound downward message.
+#[derive(Encode, Decode, Clone, Default, Copy, sp_runtime::RuntimeDebug, PartialEq, TypeInfo)]
+#[cfg_attr(feature = "std", derive(serde::Serialize, serde::Deserialize))]
+pub struct MessageIdx {
```

Comment:
nit (and below): I'd prefer `Index` suffix as it is consistent with `WrappingIndex`, `ValidatorIndex`, https://doc.rust-lang.org/std/ops/trait.IndexMut.html, etc.